dvc/output/base.py

```diff
@@ -154,10 +158,14 @@ def exists(self):
         return self.remote.exists(self.path_info)
 
     def changed_checksum(self):
+        cs = self.checksum
+        if cs is None:
+            self.remote.update_state_checksum(self.path_info)
```

Comment:
this fixes a bug while we use unit test and terminal in the same time.
if we initial the environment by test(e.g. add foo), and then change the config to `sha256` in terminal, then it shows that it will uses the hash from state in the dvc file directly without rehash(e.g. in the dvc file, out: sha256: xxxxxx(this is a md5)).

Reply:
I don't think this is a proper way to solve this. We should have a way to distinguish between different checksums in the state file itself. Seems like introducing a column for each checksum type would be a good way to go.

Reply:
It is a quick hack.
if cs is None, means config `hash.local` has been changed and no record fits, same to state record, so this update to state is needed anyway. 
Introducing a column for each checksum type, many places need to be changed.

Reply:
For example in the unit test:
```python
    def test_incompatibility(self):
        ret = main(["add", self.FOO])
        self.assertEqual(0, ret)

        ret = main(["config", "hash.local", "sha256"])
        self.assertEqual(0, ret)

        ret = main(["status"])
        self.assertEqual(0, ret)

        ret = main(["status", "-q"])
        self.assertEqual(1, ret)

        ret = main(["add", self.FOO])
        self.assertEqual(0, ret)
```

Reply:
`quick hack` is not something we should merge :slightly_smiling_face: We need a proper solution.

Not sure how that unit test is related to state db.

Reply:
FYI, the unit test is just to show why and how this `update_state_checksum` exists.

Just curious, checksums are passed to state to save, why can not be called by other to update while the checksum is not valid for new checksum type after changing the hash configuration without compatibility?

If introducing checksum types to state to save or other ways for distinguishing between different checksums in the state file itself, means all related places have to be refactored, e.g. CURD operations, implement validate, etc.

For example to introduce checksum types to state to save or other ways for distinguishing between different checksums in the state file itself:
 
Pseudo steps about getting checksum from state.py State.get(self, path_info) in this case:
1. Get the configuration of hash
2. Check the checksum, checksum type, then find out it is invalid
3. Update checksum and checksum type

Or just call by other to update, caller chain still does not change, output->remote->state, while other ways may change or introduce new caller to the chain.

Reply:
That is a great point! Indeed, creating columns for each additional checksum type is a bad solution. On the other hand, we still might want to keep a few checksum for each file. So currently we store `md5` in `md5` column, how about we replace it with `checksum` column instead, that would store a dict(e.g. serilized as json) instead that looks smth like `{"md5": "79054025255fb1a26e4bc422aef54eb4", "sha256": "b8423365e1e2245851689f675dab11574532c71bfe2e3e53aa10fa0158e28e2b"}`? It is the same way we treat checksum dicts in other places in the code.

Reply:
Why is keeping `md5` there, after config was changed to `sha256`?
What is this `state` designed for?

Reply:
@vyloy State db is used to optimize checksum calculation. It saves file checksum so that you don't have to re-compute it again later if file itself didn't change (based on mtime, size, inode). State db is used internally by each remote that needs it and when they need it. The logic that you are introducing is exposing it outside. Also, you are deciding to update state entries based on whether or not needed checksum is listed in the dvc file for this output, which is fragile.

Reply:
from output/base.py
```python
    def changed_checksum(self):
        cs = self.checksum
        if cs is None:
            self.remote.update_state_checksum(self.path_info)
            return True
        return (
            // get_checksum called in save_info has the exactly same logic with update_state_checksum does
            cs
            != self.remote.save_info(self.path_info)[ 
                self.remote.get_prefer_hash_type()
            ]
        )
```
`get_checksum` called in `save_info` has the exactly same logic with `update_state_checksum` does...

Reply:
State db is working like a cache, why can not tell a cache something is out of date?
In website developments, if used redis as a cache, redis has to find all the values are out of date by itself?

Reply:
@vyloy

> get_checksum called in save_info has the exactly same logic with update_state_checksum does...

Notion of state is not used anywhere in outputs though, because they shouldn't care about it, it is the headache of the underlying remote and you are exposing it outside.

> State db is working like a cache, why can not tell a cache something is out of date?
In website developments, if used redis as a cache, redis has to find all the values are out of date by itself?

We can, but not this way :) The fact that this particular output doesn't have some checksum doesn't mean that it is not used anywhere else.

Reply:
I still don't get your point...
Why `self.remote.save_info` can expose while `self.remote.update_state_checksum` can not?

For example?
> The fact that this particular output doesn't have some checksum doesn't mean that it is not used anywhere else.

Reply:
> Why self.remote.save_info can expose while self.remote.update_state_checksum can not?

Because `save_info` is not exposing state notion, while `update_state_checksum` does. Your `update_state_checksum` is almost a full copy of `get_checksum` and it is called right before it, which is just not right. What we could do instead is leave checksum type notion inside of the Remote, so it is the one dealing with it without you having to call dirty hacks from Output. For example, you could get rid of `update_state_checksum` if you simply make `Remote.get_checksum` do smth like `checksum = self.state.get(path_info, checksum_type=self.checksum)`, so if it say doesn't have sha256, then it would return None. This would help us avoid a lot of code duplication.

> For example?

For example, you have two branches: one is using sha256 and the other old one is still using md5s. We could also decide to allow per-output/dependency checksum type configuration in the future, so the same file could be used with different checksums. So it makes sense to make this generic enough.

Reply:
If the name is exposing state notion, how about change the name to `refresh_info`? 

It will return True then, why do you say so? And it is only called while the output checksum is not match with config. That is not even the usual use case.
>it is called right before it, which is just not right.

If we leave checksum type notion inside of the Remote, how would Remote know what checksum type is output using for querying from state to compare?

Actually, I have tried this before used this most simple way, this way leads to a lot of code refactoring. 🙂 Why do people always like to ignore the simplest and straightforward way?

> For example, you could get rid of update_state_checksum if you simply make Remote.get_checksum do smth like checksum = self.state.get(path_info, checksum_type=self.checksum), so if it say doesn't have sha256, then it would return None.

IMHO, There will be a lot of advantages with this little code duplication here, e.g. better performance, less cognitive complexity etc. 🙂 

What about three or more branches using two checksum types?
> For example, you have two branches...

Reply:
I will try some refactoring to implement this. 

> That is a great point! Indeed, creating columns for each additional checksum type is a bad solution. On the other hand, we still might want to keep a few checksum for each file. So currently we store md5 in md5 column, how about we replace it with checksum column instead, that would store a dict(e.g. serilized as json) instead that looks smth like {"md5": "79054025255fb1a26e4bc422aef54eb4", "sha256": "b8423365e1e2245851689f675dab11574532c71bfe2e3e53aa10fa0158e28e2b"}? It is the same way we treat checksum dicts in other places in the code.

Reply:
@vyloy

> If the name is exposing state notion, how about change the name to refresh_info?

The fact still stays the same: you are exposing things that should be handled by `save_info` inside the remote.

> It will return True then, why do you say so? And it is only called while the output checksum is not match with config. That is not even the usual use case.

Deciding whether or not state contains something valid and needs or doesn't need a refresh based on the config is pretty fragile and will only work while everyone is accessing it through `changed_checksum`, which might not always be the case. It is the remote itself who should decide to refresh anything if it needs to, otherwise, we are breaking the non-formal API of the Remote.

> If we leave checksum type notion inside of the Remote, how would Remote know what checksum type is output using for querying from state to compare?

By using `self.remote.changed(self.path_info, self.info)` in `output.changed_checksum()` instead of comparing self.info with self.remote.save_info by-hand :slightly_smiling_face: 

> Actually, I have tried this before used this most simple way, this way leads to a lot of code refactoring. slightly_smiling_face Why do people always like to ignore the simplest and straightforward way?

Because "the simplest and straightforward" doesn't always mean "the best" :slightly_smiling_face: We are playing a long game here, so we need the design to make sense not only now but also in the future. 

> I will try some refactoring to implement this.

:+1: Thanks! Looking forward to that! :slightly_smiling_face: